src/main/scala/rise/core/equality.scala

```diff
@@ -0,0 +1,183 @@
+package rise.core
+
+import rise.core.semantics._
+import rise.core.types._
+import util.PatternMatching
+
+object equality {
+  case class Env[T](unwrap: Map[T, T]) {
+    def add(x: T, y: T): Env[T] = Env(unwrap.updated(x, y))
+    def check(x: T, y: T): Boolean = x == y || unwrap.get(x).contains(y)
+  }
+  object Env {
+    def apply[T](): Env[T] = Env(Map())
+  }
+
+  val equivNat: Env[Kind.Identifier] => Nat => Nat => Boolean = env => a => b => {
+    val natEnv = env.unwrap.collect { case (i: NatIdentifier, n: Nat) => (i, n) }
+    // substitutes elements on the left with elements on the right
+    substitute.natsInNat(natEnv, a) == b
+  }
+
+  trait TypeEq {
+    final type Eq[K <: Kind] = K#T => K#T => Boolean
+    def apply[K <: Kind] : Eq[K] = equiv[K](Env())
+    def equiv[K <: Kind] : Env[Kind.Identifier] => Eq[K]
+    def hash[K <: Kind] : K#T => Int
+  }
+
+  object typeErasure extends TypeEq {
+    override def hash[K <: Kind]: K#T => Int = _ => 0
+    override def equiv[K <: Kind]: Env[Kind.Identifier] => Eq[K] = _ => _ => _ => true
+  }
+
+  object typeUnificationAlphaEq extends TypeEq {
+    override def hash[K <: Kind]: K#T => Int = _ => 0
+    override def equiv[K <: Kind]: Env[Kind.Identifier] => Eq[K] = env => a => b => (a, b) match {
+      case (a : Type, b : Type) => (a, b) match {
+        case (TypePlaceholder, _) => true
+        case (_, TypePlaceholder) => true
+        case _ => typeAlphaEq.equiv[TypeKind](env)(a)(b)
+      }
+      case _ => typeAlphaEq.equiv(env)(a)(b)
+    }
+  }
+
+  object typeAlphaEq extends TypeEq {
+    /** Alpha equivalence on types.
+      * Short circuits in the event of syntactic equality.
```

Comment:
Did some benchmarks on the test suite:
- `master`: 5min28sec
- `equality`: 3min59sec
- `equality` without the short-circuit: 4min9sec

It might be worth getting rid of the short-circuiting and go for the more straightforward behavior.